stdlib/internal/promql/join.go

```diff
@@ -326,84 +326,74 @@ func (c *mergeJoinCache) joinTables(a, b flux.Table) (flux.Table, error) {
 	ar := make([]flux.ColReader, 0)
 	br := make([]flux.ColReader, 0)
 
+	// Release column readers after joining
+	defer a.Done()
+	defer b.Done()
+
 	// Retain column readers for joining
 	if err := a.Do(func(cr flux.ColReader) error {
 		cr.Retain()
```

Comment:
This is basically retaining all the column readers until this method returns, right?  Which kind of defeats the purpose of batching.  I think it's fine for now, but at some point we won't want to keep the entirety of both tables in memory.

Reply:
If the purpose of this is just to retain everything, then `execute.CopyTable` would be more idiomatic.

Reply:
Ah, I think I see why `CopyTable` doesn't work. @jlapacik can you add to the `BufferedTable` interface the following?

```go
Buffer(i int) ColReader
NumBuffers() int
```

That way we can reuse that code and it would be more useful.

Reply:
@wolffcm @jsternberg the only time these column readers are not present in memory is when they are coming directly from storage. Otherwise they are already in memory and retaining them only increases the reference count.

@jsternberg we would need to add these methods to the actual column reader interface would we not? Otherwise I would have to call `CopyTable` first, which again increases all the reference counts simultaneously.

Reply:
The `CopyTable` function returns a `flux.BufferedTable` interface. It's reasonably safe to update that one. If you want to avoid that for this PR so it's not mixed in with a "breaking change", I'm fine with the current implementation as long as a `TODO: update to use buffered table when buffered table exposes the buffers` or some other random piece of text.